File: leetsync/2316-count-hills-and-valleys-in-an-array/count-hills-and-valleys-in-an-array.go
func countHillValley(nums []int) int {
	n, count, left, right := len(nums), 0, 0, 0
	for right+1 < n {
		isValley := nums[left] > nums[right] && nums[right] < nums[right+1]
		isHill := nums[left] < nums[right] && nums[right] > nums[right+1]

		if isValley || isHill {
			count++
			left = right
		}
		right++
	}
	return count
}

File: leetsync/2316-count-hills-and-valleys-in-an-array/count-hills-and-valleys-in-an-array.java
class Solution {
    public int countHillValley(int[] nums) {
        int count = 0;
        int left = 0, right = 0;
        int n = nums.length;
        while (right + 1 < n) {
            if ((nums[left] > nums[right] && nums[right] < nums[right + 1])
                    || ((nums[left] < nums[right] && nums[right] > nums[right + 1]))) {
                count++;
                left = right;
            }
            right++;
        }
        return count;
    }
}